backend/telemetry_shard/src/main.rs

```diff
@@ -212,46 +230,84 @@ where
+    // If the outer loop ends, it fires a msg on `close_connection_rx` to ensure this ends too.
+    let (ws_tx_atomic, mut ws_rx_atomic) = futures::channel::mpsc::unbounded();
+    tokio::task::spawn(async move {
+        loop {
+            let mut bytes = Vec::new();
+            tokio::select! {
+                // The close channel has fired, so end the loop. `ws_recv.receive_data` is
+                // *not* cancel safe, but since we're closing the connection we don't care.
+                _ = close_connection_rx.recv_async() => {
+                    log::info!("connection to {:?} being closed", real_addr);
+                    break
+                },
+                // Receive data and relay it on to our main select loop below.
+                msg_info = ws_recv.receive_data(&mut bytes) => {
+                    if let Err(soketto::connection::Error::Closed) = msg_info {
+                        break;
+                    }
+                    if let Err(e) = msg_info {
+                        log::error!("Shutting down websocket connection: Failed to receive data: {}", e);
+                        break;
+                    }
+                    if ws_tx_atomic.unbounded_send(bytes).is_err() {
+                        // The other end closed; end this loop.
+                        break;
+                    }
+                }
+            }
+        }
+    });
+
+    // A periodic interval to check for stale nodes.
+    let mut stale_interval = tokio::time::interval(stale_node_timeout / 2);
+
+    // Our main select loop atomically receives and handles telemetry messages from the node,
+    // and periodically checks for stale connections to keep our ndoe state tidy.
     loop {
-        let mut bytes = Vec::new();
         tokio::select! {
-            // The close channel has fired, so end the loop. `ws_recv.receive_data` is
-            // *not* cancel safe, but since we're closing the connection we don't care.
-            _ = close_connection_rx.recv_async() => {
-                log::info!("connection to {:?} being closed by aggregator", real_addr);
-                break
-            },
-            // A message was received; handle it:
-            msg_info = ws_recv.receive_data(&mut bytes) => {
-                // Handle the socket closing, or errors receiving the message.
-                if let Err(soketto::connection::Error::Closed) = msg_info {
-                    break;
+            // We periodically check for stale message IDs and remove nodes associated with
+            // them, to prevent a buildup. We boot the whole connection if no interpretable
+            // messages have been sent at all in the time period.
+            _ = stale_interval.tick() => {
+                let stale_ids: Vec<NodeMessageId> = allowed_message_ids.iter()
+                    .filter(|(_, last_seen)| last_seen.elapsed() > stale_node_timeout)
+                    .map(|(&id, _)| id)
+                    .collect();
```

Comment:
Do we really need to `collect` the ids or can we do all the work while we iterate through them?

Reply:
Yeahh, we need to collect them at the mo because we can't remove them from the `allowed_message_ids` while iterating. There is a lovely function I wanted to use called `drain_filter` that would have let me iterate over IDs being removed and do something with them, but it's not stable yet!